### PR TITLE
[FLINK-5921] [table] Add custom data types for rowtime and proctime.

### PR DESCRIPTION
**Pull request by fhueske**
We add a custom data type for the result of the `rowtime()` and `proctime()` marker functions to easier identify the chosen time semantics in when translating window operations.

Both method return a constant timestamp (0L) which is injected during constant expression reduction.

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/FunctionGenerator.scala [hunk 1]
```diff
@@ -327,6 +329,15 @@ object FunctionGenerator {
         )
       )
 
+    // generate a constant for time indicator functions.
+    //   this is a temporary solution and will be removed when FLINK-5884 is implemented.
+    case ProcTimeExtractor | EventTimeExtractor =>
+      Some(new CallGenerator {
+        override def generate(codeGenerator: CodeGenerator, operands: Seq[GeneratedExpression]) = {
+          GeneratedExpression("0L", "false", "", LONG_TYPE_INFO)
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/FunctionGenerator.scala]: Can we change `LONG_TYPE_INFO` to `SqlTimeTypeInfo.TIMESTAMP`?

Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/FunctionGenerator.scala]: Yes, thanks for the hint

**Comment by fhueske** [comment]
Updated

**Comment by sunjincheng121** [comment]
@fhueske Thanks for updating, the PR looks very good to me. I look forward to merge.
Best,
SunJincheng

**Comment by twalthr** [comment]
I'm fine with this temporary solution. +1 for merging until FLINK-5884 is fixed.

**Comment by fhueske** [comment]
Thanks for the feedback @sunjincheng121 and @twalthr 

Merging this.

**Pull request closed without merging**